### ISSUE TITLE
Issues with prometheus exposed metrics

### ISSUE THREAD

**Issue by paulfantom** [netdata/netdata#4680]
##### Bug report summary

Data exposed currently is useless from prometheus perspective. We need to fix:
- [ ] drop timestamp
- [ ] stop doing precalculations, prometheus prefers raw data
- [ ] counters are better than absolute values
- [ ] try to remove `honor_labels: true` if possible

@SuperQ please chime in to identify more issues.

##### OS / Environment

All, system independent.

##### Component Name

Prometheus backend

**Comment by SuperQ** [comment]
The pre-calculations is probably the easiest quick benefit, since it will reduce the amount of duplicate data output, which will reduce scrape time/bandwidth, and storage requirements to use netdata.

**Comment by ktsaou** [comment]
ok, let me check:

1. timestamps: we can just comment them out.
2. I am not sure I understand what do you mean with `precalculations`: when using `as-collected` mode, netdata does not process the collected values in any way. It exposes the last value collected, exactly as-collected.
3. counters vs gauges: when using `as-collected` mode, everything is sent in the format it is collected (counters as counters and gauges as gauges).
4. `honor_labels: true`: I am not sure what this is. Can you elaborate a bit?

I understand 2 and 3 are about the default `average` mode in netdata. You probably suggest to switch the default to `as-collected`. Unfortunately this will be a breaking change for netdata so we will have to switch it after a major release. Keep in mind it is very easy to configure the one needed at the netdata URL (just append `&data=as-collected` or `&data=raw`).

**Comment by SuperQ** [comment]
Ahh, thanks for the explanation. Yes, because Prometheus is a monitoring system itself, it expects as-collected or raw data. Can you explain the differences between `as-collected` and `raw`?

**Comment by ktsaou** [comment]
> Can you explain the differences between `as-collected` and `raw`?

Just aliases. Both do the same thing: expose the latest value of the collected metrics, exactly as they were collected.

So, netdata still collects all the metrics every second (for its own purposes), but keeps a copy of the latest collected value, in raw format - as it was collected.

Adding `data=as-collected` or `data=raw` instructs the prometheus exporter inside netdata, to ignore the netdata database entirely and just expose the latest collected value, as it was collected by the data collection plugins.

### Regarding the timestamp

Netdata keeps the exact timestamp each value was collected, in nanoseconds (`struct timeval`). It does this per data collection source, so if we collect 10 metrics from a single source in one go, all these metrics will share the same timestamp.

Generally, this timestamp is quite important if you plan to scrape netdata frequently.

On busy cloud environments, data collection jitter randomly reaches even 100ms. Check this latency graph (from a production cloud server):

![peek 2018-09-08 14-22](https://user-images.githubusercontent.com/2662304/48938956-699f4080-ef1b-11e8-8809-369ea1a298b9.gif)

Without taking into account this timestamp, the error rate can be 10% if you scrape netdata, from prometheus, every second.

Of course if you scrape netdata every 10s, this error smooths out to 1%.

If you collect metrics as `average`, this issue is not propagated to prometheus. Netdata does interpolation using these timings and will provide very accurate metrics no matter when they are scraped by prometheus. But, all metrics will be gauges...

**Comment by ktsaou** [comment]
Can you explain what `honor_labels: true` is and what should netdata do?

**Comment by SuperQ** [comment]
Prometheus works similarly, we collect the raw data in one shot, and timestamp it (millisecond resolution). This is why we prefer counters, as it eliminates the sampling jitter issues with gauges.

In the case of scraping netdata, we may want to keep the timestamp, as we can honor it as an authoritative TSDB source. We usually want targets to not include timestamps, as we're trying to track the observation. But in the case of netdata, it's already done the observation and we're just federating the data in.

`honor_labels` allows Prometheus to accept the `instance` and `job` labels coming from the metrics source. If we're going to treat netdata as basically a federation target, we don't need to change this.

I think we just need to switch the default to `raw` data mode.

**Comment by ktsaou** [comment]
> I think we just need to switch the default to `raw` data mode.

ok, the next major release will have this change. Until then, users will have to add it to the URL.

> Prometheus works similarly, we collect the raw data in one shot, and timestamp it (millisecond resolution).

This could be an issue for exporters that expose many sources (like node exporter). There could be a significant delay between the first and the last metric collected.

How do you solve this with counters, without taking into account the timestamp each of the metrics was collected?

**Comment by SuperQ** [comment]
@ktarasz The node_exporter is a bit of an exception, on the high side of time-per-gathering. Even still, a typical collection happens in under 15 milliseconds. That timing includes full round-trip-time, I think the gathering internally happens in under 10ms, and we haven't done any parallel operations to reduce this.

On the typical scrape interval of 15 seconds, we're talking about less than 0.1% error in time-stamping. Even at 1 second scraping, it's still less than 1% error in timing.

This is part of the reason we advocate for metrics coming from components, as closely as possible, rather than one-big-agent. We eliminate many of the problems, including timing, by talking directly to apps.

I guess what I meant by counter improving things, I was talking about information loss with gauges that expose averages. Not the error introduced by collection timing jitter. My mistake.

**Comment by ktsaou** [comment]
ok.

In netdata things are bit more optimal (due to optimized C code). So, a typical data collection is about 5ms for the whole proc.plugin (includes almost everything netdata collects from the system).

Still the problem is that on busy cloud servers this gives latency spikes up to 20x, not for the whole proc.plugin but for random parts of it. I traced these and found they are blocking I/O while reading `/proc`. This is why I decided to chart them.

I guess node.exporter has similar spikes too. The problem with a shared timestamp is that the whole data collection iteration is significantly wrong, just because one `/proc` file blocked for 100ms.

Hopefully, since you will honor netdata timestamps, this will not be the case for netdata.

> This is part of the reason we advocate for metrics coming from components, as closely as possible, rather than one-big-agent. We eliminate many of the problems, including timing, by talking directly to apps.

Yes, I agree. In netdata each module (i.e component of a plugin) has its own timings and these timings are exposed to prometheus.

> I guess what I meant by counter improving things, I was talking about information loss with gauges that expose averages.

In the case of netdata, there is no information loss when using `average`. In this mode, netdata remembers when prometheus last collected metrics and provides the average since the last data collection by the specific prometheus instance (it uses the client IP by default, but you can provide a name for it, using `&server=NAME` at the URL). So, there are no gaps or overlaps in the `average` values, even if prometheus varies its data collection frequency.

Example:

```sh
curl -Ss 'http://registry.my-netdata.io/api/v1/allmetrics?format=prometheus&help=true&server=test1' 2>/dev/null| head -n 1
# COMMENT netdata "london3.my-netdata.io" to prometheus "test1", source "average", last seen 2 seconds ago, time range 1543006399 to 1543006401
```

If you repeatedly call the above, you will see that netdata always responds with all the metrics since the last prometheus iteration. Change `&server=test1` to something else to check it.

In my tests, `average` mode is far more accurate for high resolution data collection (1s granularity), because netdata interpolates all the metrics to second-boundaries, so the average is always perfect (accuracy is 0.00001%).

Anyway, you know best what you need in prometheus. So, we will switch the default to `as-collected` at the next major release of netdata. Until then, users will be instructed to use `as-collected` with prometheus.

**Comment by ktsaou** [comment]
@paulfantom 	if there is not anything more on this, I think we can close it.

@cakrit you are maintaining a list of blocking changes to be introduced in v2 of netdata, I think related to #807 . So, let's add this change to this list, and close this issue.

**Comment by SuperQ** [comment]
Ahh, it wasn't obvious by your description that netdata had some kind of internal tracking. Is this being tracked by client IP? It sounds fragile.

**Comment by ktsaou** [comment]
> Is this being tracked by client IP? It sounds fragile.

IP by default, but each prometheus can name itself at the URL with `&server=NAME`. So, it is up to configuration.

**Comment by SuperQ** [comment]
Sounds reasonable, we should make sure this is documented for users who want to use the averages.

It turns out I was mistaken. We do collect all data in the node_exporter in parallel using a [goroutine sync group](https://github.com/prometheus/node_exporter/blob/9c9e17aba7d1995dd8a1abdcece42c04f4af4095/collector/collector.go#L112-L123). This should nicely align the data coming out of the exporter, so problems like [this](https://github.com/prometheus/node_exporter/issues/1086) won't cascade into other collectors. (My laptop takes 160ms to collect CPU frequency due to the kernel change, Ugh)

**Comment by SuperQ** [comment]
We converted our demo to use `data=raw`. The data [looks to be correct now](https://prometheus.demo.cloudalchemy.org/graph?g0.range_input=6h&g0.step_input=60&g0.expr=sum%20without%20(dimension%2Cchart)%20(irate(netdata_cpu_cpu_total%7Bdimension!%3D%22idle%22%7D%5B4m%5D))&g0.tab=0).

**Comment by ktsaou** [comment]
nice, thank you!
